tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java

```diff
@@ -0,0 +1,522 @@
+                log.error("Task {} got unexpected worker completion from {} while " +
+                    "in {} state.", id, nodeName, task.state);
+                return null;
+            }
+            boolean broadcastStop = false;
+            if (task.state == ManagedTaskState.RUNNING) {
+                task.state = ManagedTaskState.STOPPING;
+                broadcastStop = true;
+            }
+            task.maybeSetError(error);
+            task.activeWorkers.remove(nodeName);
+            if (task.activeWorkers.size() == 0) {
+                task.doneMs = time.milliseconds();
+                task.state = ManagedTaskState.DONE;
+                log.info("Task {} is now complete on {} with error: {}", id,
+                    Utils.join(task.workers, ", "),
+                    task.error.isEmpty() ? "(none)" : task.error);
+            } else if (broadcastStop) {
+                log.info("Node {} stopped.  Stopping task {} on worker(s): {}",
+                    id, Utils.join(task.activeWorkers, ", "));
+                for (String workerName : task.activeWorkers) {
+                    nodeManagers.get(workerName).stopWorker(id);
+                }
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Get information about the tasks being managed.
+     */
+    public TasksResponse tasks() throws ExecutionException, InterruptedException {
+        return executor.submit(new GetTasksResponse()).get();
+    }
+
+    class GetTasksResponse implements Callable<TasksResponse> {
+        @Override
+        public TasksResponse call() throws Exception {
+            TreeMap<String, TaskState> states = new TreeMap<>();
+            for (ManagedTask task : tasks.values()) {
+                states.put(task.id, task.taskState());
+            }
+            return new TasksResponse(states);
+        }
+    }
+
+    public void beginShutdown(boolean stopAgents) throws ExecutionException, InterruptedException {
+        if (shutdown.compareAndSet(false, true)) {
+            executor.submit(new Shutdown(stopAgents));
+        }
+    }
+
+    public void waitForShutdown() throws ExecutionException, InterruptedException {
+        while (!executor.awaitTermination(1, TimeUnit.DAYS)) { }
+    }
+
+    class Shutdown implements Callable<Void> {
+        private final boolean stopAgents;
+
+        Shutdown(boolean stopAgents) {
+            this.stopAgents = stopAgents;
+        }
+
+        @Override
+        public Void call() throws Exception {
+            log.info("Shutting down TaskManager{}.", stopAgents ? " and agents" : "");
+            for (NodeManager nodeManager : nodeManagers.values()) {
+                nodeManager.beginShutdown(stopAgents);
+            }
+            for (NodeManager nodeManager : nodeManagers.values()) {
+                nodeManager.waitForShutdown();
+            }
+            executor.shutdownNow();
```

Comment:
Isn't this the executor that is running the `Shutdown` task? Use `executor.shutdown()`?